aopy/preproc.py

```diff
@@ -1374,3 +1374,187 @@ def proc_lfp(data_dir, files, result_dir, result_filename, overwrite=False):
+    Args:
+         data_dir (str): folder containing the data you want to load
+         files (dict): a dictionary that has 'ecube' as the key
+    
+    Returns:
+        eye_data(np.ndarray): number of time points by number of number of channels
+        eye_metadata(dict): metadata when accessing eye data
+    """
+    
+    eye_metadata = dict()
+    
+    bmi3d_events, bmi3d_event_metadata = load_bmi3d_hdf_table(data_dir, files['hdf'], 'sync_events')
+
+    # get eye channels 
+    if 'left_eye_ach' in bmi3d_event_metadata and 'right_eye_ach' in bmi3d_event_metadata:
+        eye_channels = bmi3d_event_metadata['left_eye_ach'] + bmi3d_event_metadata['right_eye_ach']
+        if debug: print(f'use bmi3d supplied eye channel definition {eye_channels}')
+    else:
+        eye_channels = [8, 9, 10, 11]
+        if debug: print(f'eye channel definitions do not exist, use eye channels {eye_channels} ')
+        
+    eye_metadata['eye_channels'] = eye_channels
+    eye_metadata['left_eye_ach_proc']  = [0,1]
+    eye_metadata['right_eye_ach_proc']  = [2,3]
+    
+    # get eye data
+    analog_data, analog_metadata = load_ecube_analog(data_dir, files['ecube'], channels=eye_channels)
+    eye_metadata.update(analog_metadata)
+    
+    #scale eye data from bits to volts
+    if 'voltsperbit' in analog_metadata:
+        analog_voltsperbit = analog_metadata['voltsperbit']
+    else:
+        analog_voltsperbit = 3.0517578125e-4
+        eye_metadata['voltsperbit'] = analog_voltsperbit
+        
+    eye_data = analog_data * analog_voltsperbit
+    
+    
+    return eye_data, eye_metadata
+
+
+def calc_eye_calibration(exp_data, exp_metadata, debug=True):
+    """
+    extracts cursor data and eyedata and caliberate, align them and calculate the least square fitting coefficients
+    
+    Args:
+        exp_data(np.ndarray): parsed bmi3d data with 'eye_data' as an variable
+        exp_metadata(dict): 
+    """
+    #get cursor kinematics
+    cursor_kinematics = exp_data['task']['cursor'][:,[0,2,1]] # cursor (x, z, y) position on each bmi3d cycle
+    events = exp_data['events']
+    eye_data = exp_data['eye_data']
+    eye_metadata = exp_metadata['eye_metadata']
+    
+    aligned_cursor, eye_data_aligned = align_eye_data(events, cursor_kinematics, eye_data, debug = debug)
+    
+    # caliberate each eye respectively
+    le_aligned = eye_data_aligned[:, exp_metadata['eye_metadata']['left_eye_ach_proc']]
+    re_aligned = eye_data_aligned[:, exp_metadata['eye_metadata']['right_eye_ach_proc']]
+    
+    left_eye_caliberation = calc_single_eye_caliberation(aligned_cursor, le_aligned)
+    right_eye_caliberation = calc_single_eye_caliberation(aligned_cursor, re_aligned)
+    
+    caliberation_file = {
+        'left_eye':left_eye_caliberation,
+        'right_eye':right_eye_caliberation
+    }
+    
+    return caliberation_file
+    
+    
+
+    
+def calc_single_eye_caliberation(cursor_trajectories_aligned, eye_aligned):
```

Comment:
could you make this into a more general function and put it into analysis.py? basically it takes two 2D trajectories and computes the linear regression between their x and y coordinates. doesn't have to be eye and cursor data, necessarily.

since you're having issues with the save_hdf() and dictionaries of dictionaries, maybe it would be better to return the x and y coefficients together. something like 

```
coefficients = np.array([[slope_x, intercept_x], [slope_y, intercept_y]]),
correlation_coeff = np.array([r_value_x, r_value_y])
```

then you could save four variables into the hdf file: left_eye_coefficients, right_eye_coefficients, left_eye_correlation_coeff, right_eye_correlation_coeff

Reply:
it should save and added caliberate funcs, 
should work, 
need to test on a few more days of  data.